File: array/leetcode/26/c/sol_2.c
/**
 * @file sol_2.c
 * @author LiYu87 (mickey9910326@gmail.com)
 * @brief leetcode 26. Remove Duplicates from Sorted Array
 * @date 2020.12.09
 * 
 * Remove variable "val", and use "idx" as a index of 
 * new array to get previous value.
 * And get new array count by "idx" + 1 .
 */

int removeDuplicates(int* nums, int numsSize) {
    // exception for null array []
    if (numsSize == 0) {
        return 0;
    }
    
    int idx = 0;  ///< ndex of new array

    /**
     * Index 0 of intput array is always a new item for new array.
     * And it's already in the index 0 of the array.
     * Dosent need to assign value again.
     */

    for (int i = 1; i < numsSize; i++) {
        if (nums[i] != nums[idx]) {
            nums[idx] = nums[i];
        }
    }

    return idx + 1;
}

File: array/leetcode/26/py/sol_2.py
"""leetcode 26. Remove Duplicates from Sorted Array
sol.2 fix line21
Only use [1::] of list to compare, index 0 is used im line 15 and 16.
"""

class Solution:
    def removeDuplicates(self, nums: List[int]) -> int:
        #  exception for null array []
        if not nums:
            return 0
        elif len(nums) == 1:
            # this condiction cost less than "for num in nums[1::]"
            return 1
        
        count = 1  # count of new array items
        val = nums[0] # previous value to compare in array search

        # Index 0 of intput array is always a new item for new array.
        # And it's already in the index 0 of the array.
        # Dosent need to assign value again.

        for num in nums[1::]:
            if num != val:
                val = num
                nums[count] = val
                count = count + 1
        
        return count
